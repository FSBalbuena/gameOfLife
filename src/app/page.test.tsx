import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import GameOfLife from "@/app/page";

describe("GameOfLife", () => {
  describe("renders", () => {
    it('User should be able to see the title "Conway\'s game of life"', () => {
      render(<GameOfLife />);

      const heading = screen.getByRole("heading", { level: 1 });

      expect(heading).toBeInTheDocument();
      expect(heading).toHaveTextContent("Conway's Game of Life");
    });

    it('User should be able to see a description "A cellular automaton simulation."', () => {
      render(<GameOfLife />);

      const description = screen.getByRole("paragraph");

      expect(description).toBeInTheDocument();
      expect(description).toHaveTextContent("A cellular automaton simulation.");
    });
    it('User should be able to see the "Next state" button', () => {
      render(<GameOfLife />);

      const button = screen.getByRole("button", {
        name: "Next state",
      });

      expect(button).toBeInTheDocument();
      expect(button).toHaveTextContent("Next state");
    });

    it('User should be able to see the "Advance Steps"', () => {
      render(<GameOfLife />);

      const button = screen.getByRole("button", {
        name: "Advance state",
      });

      const input = screen.getByRole("spinbutton");

      expect(button).toBeInTheDocument();
      expect(input).toBeInTheDocument();
      expect(button).toHaveTextContent("Advance state");
    });

    it('User should be able to see the "Play Forever" checkbox', () => {
      render(<GameOfLife />);

      const checkbox = screen.getByRole("checkbox", {
        name: "Play Forever",
      });

      expect(checkbox).toBeInTheDocument();
    });

    it('"Play Forever" checkbox should be unchecked ', () => {
      render(<GameOfLife />);

      const checkbox = screen.getByRole("checkbox", {
        name: "Play Forever",
      });

      expect(checkbox).toBeInTheDocument();
      expect(checkbox).not.toBeChecked();
    });

    it("renders homepage unchanged", () => {
      //mostly for style changes
      //"User should be able to see the grid"
      const { container } = render(<GameOfLife />);
      expect(container).toMatchSnapshot();
    });
  });

  xdescribe("behaviour", () => {
    xdescribe("Grid", () => {
      xit("User should be able to select the initial state", () => {});
    });
    xdescribe("Controls", () => {
      xdescribe("when click on Next step", () => {
        xit('User should be able to click on "Next state" button', () => {});
        xit("User should be able to see a loading state", () => {});
        xit("User should be able to see the game's next state", () => {});
      });
      xdescribe("when checks Play Forever", () => {
        xit('User should be able to check "Play Forever" checkbox', () => {});
        xit("User should be able to see the game's next state every second", () => {});
        xit('User should be able to see a message "Automated Game" on top of the grid', () => {});
        xit("Grid should be disabled", () => {});
        xit('User should be able to uncheck "Play Forever" checkbox', () => {});
      });
      xdescribe("when uses Advance Steps", () => {
        xit('User should be able to type a number on "Advance Steps" input', () => {});
        xit('User should be able to click on "Advance steps" button.', () => {});
        xit("User should be able to see the game's next state", () => {});
        xit("User should not be able to place a negative number, or zero.", () => {});
      });
    });
  });

  xdescribe("errors", () => {
    xdescribe("User should be able to see an error message", () => {
      xit("when he places a negative number", () => {});
      xit("when there is an internal error", () => {});
    });
    xit("User should be redirected if he places a wrong url.", () => {});
  });
});
