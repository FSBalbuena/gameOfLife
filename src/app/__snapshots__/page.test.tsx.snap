// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameOfLife renders renders homepage unchanged 1`] = `
<div>
  <main
    class="flex flex-col align-center items-center p-10 pt-20 md:w-3/5 lg:w-1/2 xl:w-2/5 mx-auto"
  >
    <h1
      class="text-4xl mb-4"
    >
      Conway's Game of Life
    </h1>
    <p>
      A cellular automaton simulation.
    </p>
    <canvas
      class="w-full h-full"
      id="game-of-life-canvas"
    />
    <label
      class="flex gap-2 p-1 self-end"
      for="play-forever"
    >
      <input
        id="play-forever"
        type="checkbox"
      />
      Play Forever
    </label>
    <div
      class="flex w-full justify-center gap-10"
      role="group"
    >
      <button
        class="fold-bold rounded border-2 border-black bg-white px-3 py-1 text-base font-bold text-black transition duration-100 hover:bg-yellow-400 hover:text-gray-900"
        type="button"
      >
        Next state
      </button>
      <div
        class="flex"
      >
        <input
          class="border-2 p-1 rounded-md rounded-r-none w-10 text-center"
          min="1"
          type="number"
        />
        <button
          class="fold-bold rounded rounded-l-none border-2 border-black bg-black px-3 py-1 text-base font-bold text-white transition duration-100 hover:bg-gray-900 hover:text-yellow-500"
          type="button"
        >
          Advance state
        </button>
      </div>
    </div>
  </main>
</div>
`;
